repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/core/wrappers/response/TomcatResponseWrapperTest.groovy

```diff
@@ -0,0 +1,101 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package features.core.wrappers.response
+
+import org.openrepose.framework.test.ReposeConfigurationProvider
+import org.openrepose.framework.test.ReposeContainerLauncher
+import org.openrepose.framework.test.TestProperties
+import org.rackspace.deproxy.Deproxy
+import spock.lang.Shared
+import spock.lang.Specification
+
+class TomcatResponseWrapperTest extends Specification {
+
+    static final int ERROR_CODE = 419
+    static final String ERROR_HEADER_NAME = "Error"
+    static final String ERROR_MESSAGE = "OHJEEZ"
+    static final String ERROR_BODY = "Your request is bad and you should feel bad!"
+
+    @Shared
+    Deproxy deproxy
+
+    @Shared
+    ReposeContainerLauncher tomcatLauncher
+
+    @Shared
+    TestProperties properties = new TestProperties(this.getClass().canonicalName.replace('.', '/'))
+
+    def setupSpec() {
+        def configurationProvider = new ReposeConfigurationProvider(properties)
+
+        def params = properties.defaultTemplateParams
+        params += [
+            clusterId   : 'cluster1',
+            nodeId      : 'node1',
+            errorCode   : ERROR_CODE,
+            errorMessage: ERROR_MESSAGE,
+            errorBody   : ERROR_BODY,
+            headerName  : ERROR_HEADER_NAME
+        ]
+        configurationProvider.applyConfigs("common", params)
+        configurationProvider.applyConfigs("features/core/wrappers/response/war", params)
+
+        deproxy = new Deproxy()
+        deproxy.addEndpoint(properties.targetPort)
+
+        tomcatLauncher = new ReposeContainerLauncher(
+            configurationProvider,
+            properties.tomcatJar,
+            params.clusterId as String,
+            params.nodeId as String,
+            properties.reposeRootWar,
+            properties.reposePort)
+        tomcatLauncher.start()
+        tomcatLauncher.waitForDesiredResponseCodeFromUrl(reposeEndpoint, 200)
+    }
+
+    def cleanupSpec() {
+        deproxy?.shutdown()
+        tomcatLauncher?.stop()
+    }
+
+    def "a filter writes a header and response body then calls sendError on the response"() {
+        when: "a request is made to Repose"
+        def messageChain = deproxy.makeRequest(reposeEndpoint + "/scripting")
+
+        then: "the response code should be the error code, not a 500"
+        messageChain.receivedResponse.code as Integer == ERROR_CODE
+
+        // TODO: For some reason, Tomcat is not sending any reason phrase in this case
```

Comment:
I don't know that it is strictly part of the spec, so maybe they chose to throw it away?

Reply:
Well that would explain not writing _our_ reason phrase, but it seems very odd to me that _no_ reason phrase is written.

Apparently, however, that may be allowable by HTTP specification. The relevant ABNF is:
```
Status-Line = HTTP-Version SP Status-Code SP Reason-Phrase CRLF
Status-Code    =
            "100"  ; Section 10.1.1: Continue
          | "101"  ; Section 10.1.2: Switching Protocols
          | "200"  ; Section 10.2.1: OK
          | "201"  ; Section 10.2.2: Created
          | "202"  ; Section 10.2.3: Accepted
          | "203"  ; Section 10.2.4: Non-Authoritative Information
          | "204"  ; Section 10.2.5: No Content
          | "205"  ; Section 10.2.6: Reset Content
          | "206"  ; Section 10.2.7: Partial Content
          | "300"  ; Section 10.3.1: Multiple Choices
          | "301"  ; Section 10.3.2: Moved Permanently
          | "302"  ; Section 10.3.3: Found
          | "303"  ; Section 10.3.4: See Other
          | "304"  ; Section 10.3.5: Not Modified
          | "305"  ; Section 10.3.6: Use Proxy
          | "307"  ; Section 10.3.8: Temporary Redirect
          | "400"  ; Section 10.4.1: Bad Request
          | "401"  ; Section 10.4.2: Unauthorized
          | "402"  ; Section 10.4.3: Payment Required
          | "403"  ; Section 10.4.4: Forbidden
          | "404"  ; Section 10.4.5: Not Found
          | "405"  ; Section 10.4.6: Method Not Allowed
          | "406"  ; Section 10.4.7: Not Acceptable
          | "407"  ; Section 10.4.8: Proxy Authentication Required
          | "408"  ; Section 10.4.9: Request Time-out
          | "409"  ; Section 10.4.10: Conflict
          | "410"  ; Section 10.4.11: Gone
          | "411"  ; Section 10.4.12: Length Required
          | "412"  ; Section 10.4.13: Precondition Failed
          | "413"  ; Section 10.4.14: Request Entity Too Large
          | "414"  ; Section 10.4.15: Request-URI Too Large
          | "415"  ; Section 10.4.16: Unsupported Media Type
          | "416"  ; Section 10.4.17: Requested range not satisfiable
          | "417"  ; Section 10.4.18: Expectation Failed
          | "500"  ; Section 10.5.1: Internal Server Error
          | "501"  ; Section 10.5.2: Not Implemented
          | "502"  ; Section 10.5.3: Bad Gateway
          | "503"  ; Section 10.5.4: Service Unavailable
          | "504"  ; Section 10.5.5: Gateway Time-out
          | "505"  ; Section 10.5.6: HTTP Version not supported
          | extension-code
      extension-code = 3DIGIT
      Reason-Phrase  = *<TEXT, excluding CR, LF>
```

Reply:
Well anyway, since this is known behavior, I'll update this test to reflect that and drop the TODO, at least.